stan/math/prim/mat/prob/multi_normal_cholesky_lpdf.hpp

```diff
@@ -101,37 +102,63 @@ typename return_type<T_y, T_loc, T_covar>::type multi_normal_cholesky_lpdf(
     check_not_nan(function, "Random variable", y_vec[i]);
   }
 
+  operands_and_partials<T_y, T_loc, T_covar> ops_partials(y, mu, L);
+
   if (size_y == 0)
-    return lp;
+    return ops_partials.build(0.0);
 
   if (include_summand<propto>::value)
-    lp += NEG_LOG_SQRT_TWO_PI * size_y * size_vec;
+    logp += NEG_LOG_SQRT_TWO_PI * size_y * size_vec;
+
+  matrix_partials_t L_dbl = value_of(L);
+  // do a single inversion
+  matrix_partials_t inv_trans_L_dbl
+      = mdivide_left_tri<Eigen::Upper>(transpose(L_dbl));
 
-  if (include_summand<propto, T_covar_elem>::value)
-    lp -= L.diagonal().array().log().sum() * size_vec;
+  // analytic expressions taken from
+  // http://qwone.com/~jason/writing/multivariateNormal.pdf
+  // written by Jason D. M. Rennie
+  // expressions adapted to avoid (most) inversions
 
   if (include_summand<propto, T_y, T_loc, T_covar_elem>::value) {
-    lp_type sum_lp_vec(0.0);
     for (size_t i = 0; i < size_vec; i++) {
-      Eigen::Matrix<typename return_type<T_y, T_loc>::type, Eigen::Dynamic, 1>
-          y_minus_mu(size_y);
+      vector_partials_t y_minus_mu_dbl(size_y);
       for (int j = 0; j < size_y; j++)
-        y_minus_mu(j) = y_vec[i](j) - mu_vec[i](j);
-      Eigen::Matrix<typename return_type<T_y, T_loc, T_covar>::type,
-                    Eigen::Dynamic, 1>
-          half(mdivide_left_tri_low(L, y_minus_mu));
-      // FIXME: this code does not compile. revert after fixing subtract()
-      // Eigen::Matrix<typename
-      //               boost::math::tools::promote_args<T_covar,
-      //                 typename value_type<T_loc>::type,
-      //                 typename value_type<T_y>::type>::type>::type,
-      //               Eigen::Dynamic, 1>
-      //   half(mdivide_left_tri_low(L, subtract(y, mu)));
-      sum_lp_vec += dot_self(half);
+        y_minus_mu_dbl(j) = value_of(y_vec[i](j)) - value_of(mu_vec[i](j));
+
+      vector_partials_t half
+          = mdivide_left_tri<Eigen::Lower>(L_dbl, y_minus_mu_dbl);
+      // alternative which avoids inversions
```

Comment:
I don't think we need the inverse for a determinant.  I think what we want is a solver where we can grab solutions (for division) and also the determinant.

Reply:
I am not sure if I follow here what you are saying. So we need the inverse of L in line 157 and hence I decided to calculate it.

The alternative (from my understanding) to what I have put into the code is to use in line 134 a mdivide_left_tri call to get the scaled_diff term there and then calculate the inverse of L in line 157. I see these two options... maybe if you could explain again or we discuss shortly on Thursday?

Reply:
I don't think Eigen facilitates constructing a solver object from a Cholesky factor.